runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,320 @@
+local util = require('vim.lsp.util')
+local log = require('vim.lsp.log')
+local ms = require('vim.lsp.protocol').Methods
+local api = vim.api
+
+local M = {}
+
+---@class (private) vim.lsp.folding_range.BufState
+---
+---@field version? integer
+---
+--- Index in the form of row -> foldlevel
+---@field row_level table<integer, [integer, nil | ">" | "<"]>
+---
+--- Index in the form of start_row -> collapsed_text
+---@field row_text table<integer, string>
+---
+--- Index in the form of kind -> range
+---@field kind_ranges table<lsp.FoldingRangeKind, lsp.FoldingRange[]?>>
+
+---@type table<integer, vim.lsp.folding_range.BufState?>
+local bufstates = {}
+
+--- Update visible folds in the buffer.
+---@param bufnr integer
+local function foldupdate(bufnr)
+  for _, winid in ipairs(vim.fn.win_findbuf(bufnr)) do
+    local wininfo = vim.fn.getwininfo(winid)[1]
+    if wininfo and wininfo.tabnr == vim.fn.tabpagenr() then
+      if vim.wo[winid].foldmethod == 'expr' then
```

Comment:
Is this condition really necessary, especially since the expr could be something different anyway?

Reply:
`foldupdate` to be called is possible only after the `vim.lsp.foldexpr` setup, which means that when it is called, `foldexpr` has already been ensured to be `vim.lsp.foldexpr`. This is to prevent the situation where `expr` is set but `foldmethod` is not set.